### DISCUSSION TITLE
QnA뷰 질문

### DISCUSSION THREAD

**Discussion by sokft** [ttakkeun/Front_iOS#57] [Q&A]
### 1. Top 10 질문
<img width="483" alt="스크린샷 2024-07-30 오후 4 28 51" src="https://github.com/user-attachments/assets/808ca912-f731-495b-86e5-359093960b4d">

이 사진에서 Top10질문은 그냥 고정된 질문인가요? 아니면 실제로 사용자가 질문한 것들을 바탕으로 순위를 매겨서 가져오는 질문인가요?
만약 전자면, 그냥 배열에 넣고 하면 될까요?

### 2. 카테고리 별 질문
<img width="456" alt="스크린샷 2024-07-30 오후 4 55 29" src="https://github.com/user-attachments/assets/18a7e10a-0012-4eca-a13d-a597ccb465ed">

카테고리 별 질문을 보면 질문하기 버튼도 없고, 정해진 질문이 나열되는 것 같은데 Json파일 만들어서 파싱하도록 하신 이유를 알 수 있을까요?

**그래서,** 질문하기 버튼이 없는 이상 다 고정된 질문일 것 같은데 뷰모델이 필요할까요,,?

**Comment by JEONG-J** [comment]
아니 이거 왜 또 물어봄? 배열로 하지말라고 2층에서 만났을 때 얘기했는데

이 사진에서 Top10질문은 그냥 고정된 질문인가요? 아니면 실제로 사용자가 질문한 것들을 바탕으로 순위를 매겨서 가져오는 질문인가요?
만약 전자면, 그냥 배열에 넣고 하면 될까요?
-> 전자, 배열 안됨

카테고리 별 질문을 보면 질문하기 버튼도 없고, 정해진 질문이 나열되는 것 같은데 Json파일 만들어서 파싱하도록 하신 이유를 알 수 있을까요?

-> 재서용! 데이터의 지속 가능성 및 발전성 오키?

그럼 뷰모델이 왜?

추천 탭 전체를 담당하는 뷰모델이 없으면 어떻게 초기화하고 받을래? -> 해결법 찾으면 뷰모델 안 만들어도 됨

**Comment by JEONG-J** [comment]
배열로 만들 때 코드

너가 자료구조를 배웠나?? 그 가정하에 보여줄게

Struct QnA {
 var question: String
 var ask: String
}

final class DumyData {
 static let dataList: [QnA] = [
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
]

-------

위에꺼 json 버전

{
 "question" : "털이 너무 잘 빠져요",
 "ask" : "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"
}

이렇게 만약 질문과 답이 딱 하나만 그니까 갯수가 제한되어있고 적다면? 둘 중에 뭐가 제일 유지보수와 코드 가독성 데이터 복잡도 측면에서 제일 이득인게 무엇일까? 당연히 첫 번째꺼! 왜냐고? json으로 만들면 그걸 파싱해서 데이터 구조에 넣어두는 코드를 작성해야 하는데 1개 데이터를 위해 그걸 다 작성하는게 과연 효율적인 코드 작성이 될까? 절대 아니지~

근데 우리뷰는 질문 최대 갯수 최소갯수 제한이란게 존재하지 않음 즉 앱의 지속 가능성으로 따질때 어떤 질문ㄷ과 답이 게속 등장하게 될지 미지수임

그럼 이런 경우 위 코드들이 어떻게 바뀌는지 보여줄게, 단 질문과 답은 그대로 한다는 가정하에

final class DumyData {
 static let dataList: [QnA] = [
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"),
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
 QnA(question: "털이 너무 잘 빠져요", ask: "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요")
]

첫 번째 경우임! 자 이렇게 되어있을 때, 이거 가지고 뷰가 렌더링 될텐데 시간복잡도가 어떻게 나올까? 배열을 순회하면서 나 값을 처리해야하는데
근데 배열안에 두개의 데이터 값이 존재하지? question과 ask 그럼 인덱스 하나당 2개 확인한다 이렇게 순회하면서 배열 전체를 다 돌게될꺼야 물론 이걸 알고리즘 넣어서 분할정복으로 쓰레드 나누고 쪼개서 하고 머지하면? 시간복잡도는 줄어들고 빠른 데이터 처리가 되겠지? 근데 이 상황은 마찬가지로 데이터 갯수가 몇개될지 모름 진짜 100개 존재할 수 있는거고 1000개 존재할 수 있는거임

그럼 시간 복잡도 생각해봐 쉽게 설며해서 너가 교장선생님인데 학교의 학생들은 운동장에 집합시키고 1열의 1번부터 마지막 열의 마지막 번호까지 다 하나하나 이름 물어보고 학생번호 물어보면서 다 뒤지는거임 이게 빠를거 같니? 근데 너가 이걸 매주 월요일마다 한다해봐 매주 너가 머릿속으로 외웠다면 당연히 아 이 학생 누군지 알아 하고 넘어갈 수 있겠지만 모른다면 다시 물어볼거임

즉, 배열로 저장하면 뷰에 들어올 때마다 계속 데이터 넣고 초기화하고 그래야 됨 캐싱이 사라짐

자 그럼 이게 json으로 될 경우,

{
[
 "question" : "털이 너무 잘 빠져요",
 "ask" : "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"
],
[
 "question" : "털이 너무 잘 빠져요",
 "ask" : "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"
],
[
 "question" : "털이 너무 잘 빠져요",
 "ask" : "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"
],
[
 "question" : "털이 너무 잘 빠져요",
 "ask" : "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"
],
[
 "question" : "털이 너무 잘 빠져요",
 "ask" : "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"
],
[
 "question" : "털이 너무 잘 빠져요",
 "ask" : "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"
],
[
 "question" : "털이 너무 잘 빠져요",
 "ask" : "샴푸가 문제 일 확률이 높아요! 어서 병원에 가보세요"
],

}

이렇게하고 json에서 파싱을 하고 데이터에 넣어두면 당연히 파싱하는 작업에서 시간 복잡도가 걸릴꺼임 하지만 한 번 파싱하면 추가적으로 메모리에 따로 보관하기 때문에 추후에 데이터 재로딩하면서 다시 읽는 읽이 없음 초기에만 읽으면 되거든
